Source/JavaScriptCore/parser/Parser.cpp

```diff
@@ -4219,17 +4229,18 @@ template <typename TreeBuilder> TreeExpression Parser<LexerType>::parseAssignmen
         bool isArrowFunctionToken = match(ARROWFUNCTION);
         if (!lhs || isArrowFunctionToken) {
             SavePointWithError errorRestorationSavePoint = swapSavePointForError(context, savePoint);
-            bool isAsyncArrow = false;
+            const unsigned functionStart = start;
+            bool isAsync{ false };
             if (UNLIKELY(classifier.indicatesPossibleAsyncArrowFunction())) {
                 if (matchContextualKeyword(m_vm.propertyNames->async)) {
+                    isAsync = true;
                     next();
-                    isAsyncArrow = !m_lexer->hasLineTerminatorBeforeToken();
                 }
             }
             if (isArrowFunctionParameters(context)) {
                 if (wasOpenParen)
                     currentScope()->revertToPreviousUsedVariables(usedVariablesSize);
-                return parseArrowFunctionExpression(context, isAsyncArrow);
+                return parseArrowFunctionExpression(context, isAsync, functionStart);
```

Comment:
@Constellation thanks for spotting this, it's fixed now (`isAsync` is now a separate parameter). The reason is that I initially try/hacked it by looking for async keywords, and in this case the presence/absence of this keyword would (relatively) explicitly tells if it was an async function or not. Obviously it was a totally wrong assumption now as we're looking for "function start" which can be either async, get, set, whatever JS enable to start a function depending the context.

I take this opportunity to point out something a bit odd with my changes: in lot of places in the parser, there are now `functionStart` (that I added) and `functionKeywordStart` (already there). In lot of case, both are similar. And where two are different (i.e. 'async function ...', etc.) we picked the `functionStart` when building the node `functionStart` field. (the `value_or(static_cast<...` thing in your other comment).

But the reason I didn't removed 'functionKeywordStart' (yet) is because this value is used in another UnlinkedFunctionExecutable field, `m_typeProfilingStartOffset`. I would be tempted to say "we don't need this anymore, let's just use `functionStart`:

- that would make parser code simpler/less confusing (as we would remove `functionKeywordStart`
- this is a small memory footprint win as it would be possible to remove `m_typeProfilingStartOffset` field

But is this correct from profiling tools perspective? One thing that would lead me to say "that's fine" is the fact there are a bunch of function that doesn't even have the `function` keyword, but I really have no clues about actual use case of this profiling start offset field, so... Any insight would be much appreciated :-))

Reply:
I actually just pushed commit `[37b8276](https://github.com/WebKit/WebKit/pull/14704/commits/37b827636327882dce8280626fcb701227561cdb)` which is an attempt of what I'm discussing just above.

All jsc tests (including stress ones) are passing locally, and no obvious jsc related breakage on layout ones, so I'll give a try here. 

But my question above remain, not sure if what I'm doing in this commit is totally right (regarding some profiling flow) or not.